### PR TITLE
HSEARCH-3564 Type cast too strong

### PR DESCRIPTION
**Pull request by carstenartur**
https://hibernate.atlassian.net/browse/HSEARCH-3564

### CODE REVIEW

**Comment by fax4ever** [comment]
Hi @carstenartur. Thanks for the contribution.

It seems that the commit does not compile:
~~~
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/fercoli/code/hibernate-search/backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/projection/impl/ElasticsearchSearchProjection.java:[66,65] incompatible types: java.lang.Object cannot be converted to capture#1 of ?
[INFO] 1 error
~~~

Moreover, what's the purpose of the contribution?
Many thanks
Fabio

**Comment by yrodiere** [comment]
Hi @carstenartur ,

I appreciate the contribution, and some of these suggestions are legitimate clean-ups, but some of them are wrong, like the one on `new Appendable()` in the tests, or the one that breaks the build.

There may be strong reasons, beyond everyday maintenance, to perform this cleanup, such as the project not compiling in a specific IDE. If that's the case, please create a ticket on JIRA so we can discuss how to best address these problems: https://hibernate.atlassian.net/projects/HSEARCH.

In the meantime, I'll close this PR.

**Comment by carstenartur** [comment]
No problem.
missed that error, sorry, 
May have missunderstood HSEARCH-3564

@Fabio:
This is about fixes of type casts, which are too strict and casting to a more general type would be sufficient for the actual expression. Sometimes, an instance is casted to a specific subtype, but methods invoked or fields used from the casted instance are actually defined by some supertype. In that case cast to too specific type introduces an unnecessary coupling to the code and limits its extensibility.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/projection/impl/ElasticsearchSearchProjection.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ E extract(ProjectionHitMapper<?, ?> projectionHitMapper,
 	@SuppressWarnings({ "unchecked", "rawtypes" })
 	static <Z> Z transformUnsafe(ElasticsearchSearchProjection<?, Z> projection, LoadingResult<?> loadingResult,
 			Object extractedData, SearchProjectionTransformContext context) {
-		return (Z) ( (ElasticsearchSearchProjection) projection ).transform( loadingResult, extractedData, context );
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/projection/impl/ElasticsearchSearchProjection.java]: This cast is necessary to bypass generics limitations of the compiler. We can't remove it.

File: engine/src/main/java/org/hibernate/search/engine/cfg/spi/ConvertUtils.java [hunk 2]
```diff
@@ -310,7 +310,7 @@ public static BigInteger convertBigInteger(Object value) {
 					.collect( Collectors.toList() );
 		}
 		else if ( value instanceof String ) {
-			return separatorPattern.splitAsStream( (String) value )
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/engine/cfg/spi/ConvertUtils.java]: This isn't consistent with the `instanceof` above, nor with the other `convert` methods we have in the same class. I'd be in favor of leaving this as-is.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/work/impl/LuceneSearchWork.java [hunk 3]
```diff
@@ -89,7 +89,7 @@ else if ( limit == 0L ) {
 			return 0;
 		}
 		else {
-			return Math.min( (int) ( offset + limit ), reader.maxDoc() );
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/work/impl/LuceneSearchWork.java]: This cast should indeed be removed, but the new code doesn't follow the Hibernate code style. There should be a space after an opening parenthesis (missing here), and a space before a closing parenthesis.

See http://hibernate.org/search/contribute/, in particular see the code style configuration for [Eclipse](https://github.com/hibernate/hibernate-ide-codestyles/blob/master/eclipse/hibernate-code-template-search.xml) or [IntelliJ IDEA](https://github.com/hibernate/hibernate-ide-codestyles/blob/master/intellij-14/hibernate_noorm.xml).

File: integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/query/SearchQueryFetchIT.java [hunk 4]
```diff
@@ -84,7 +84,7 @@ public void fetch_noArg() {
 
 	@Test
 	public void fetch_limit() {
-		assertThat( matchAllQuery().fetch( (Integer) null ) )
```
Comment (yrodiere) [inline integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/query/SearchQueryFetchIT.java]: Code style: there should be a space before and after `null`.

File: integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/query/SearchQueryFetchIT.java [hunk 5]
```diff
@@ -152,7 +152,7 @@ public void fetchHits_noArg() {
 
 	@Test
 	public void fetchHits_limit() {
-		assertThat( matchAllQuery().fetchHits( (Integer) null ) )
```
Comment (yrodiere) [inline integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/query/SearchQueryFetchIT.java]: Code style: there should be a space before and after `null`.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/smoke/usertype/OptionalStringUserType.java [hunk 6]
```diff
@@ -59,7 +59,7 @@ public void nullSafeSet(PreparedStatement st, Object value, int index, SharedSes
 			delegateType.nullSafeSet( st, null, index, session );
 		}
 		else {
-			delegateType.nullSafeSet( st, (String) optional.get(), index, session );
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/smoke/usertype/OptionalStringUserType.java]: Code style: there should be a space before `st`.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/model/impl/HibernateOrmPathFilterFactory.java [hunk 7]
```diff
@@ -305,7 +306,7 @@ else if ( collectionValue instanceof org.hibernate.mapping.Map ) {
 				 * Do not let ORM confuse you: getKey() doesn't return the value of the map key,
 				 * but the value of the foreign key to the targeted entity...
 				 */
-				return ( (org.hibernate.mapping.Map) collectionValue ).getIndex();
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/model/impl/HibernateOrmPathFilterFactory.java]: This is inconsistent with the `instanceof` above. I would not change it.

File: util/common/src/test/java/org/hibernate/search/util/common/impl/ToStringTreeBuilderTest.java [hunk 8]
```diff
@@ -142,8 +142,8 @@ public String toString() {
 					.startObject( "emptyChild" )
 						.endObject()
 					.attribute( "appendable", new Appendable() )
-					.attribute( "appendableAsObject", (Object) new Appendable() )
-					.attribute( "nullAppendable", (Appendable) null )
```
Comment (yrodiere) [inline util/common/src/test/java/org/hibernate/search/util/common/impl/ToStringTreeBuilderTest.java]: Code style: there should be a space after an opening parenthesis and before a closing parenthesis.

File: util/common/src/test/java/org/hibernate/search/util/common/impl/ToStringTreeBuilderTest.java [hunk 9]
```diff
@@ -142,8 +142,8 @@ public String toString() {
 					.startObject( "emptyChild" )
 						.endObject()
 					.attribute( "appendable", new Appendable() )
-					.attribute( "appendableAsObject", (Object) new Appendable() )
```
Comment (yrodiere) [inline util/common/src/test/java/org/hibernate/search/util/common/impl/ToStringTreeBuilderTest.java]: This cast allows us to check that the method accepts an `Object`. I'd keep it?

**Comment by yrodiere** [comment]
Fixed the few remaining style issues, squashed, added the ticket number to the commit message, and merged to master.
Thanks!

**Pull request closed without merging**